### PR TITLE
[TOOLS-85] Sizing Import IT

### PR DESCRIPTION
**Pull request by jhonypg**

### CODE REVIEW

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import org.junit.Test;
+
+import objective.taskboard.it.SizingImportUi.SizingStepFour;
+import objective.taskboard.it.SizingImportUi.SizingStepOne;
+import objective.taskboard.it.SizingImportUi.SizingStepThree;
+import objective.taskboard.it.SizingImportUi.SizingStepTwo;
+
+public class SizingImportIT extends AuthenticatedIntegrationTest {
+	
+	final static String SPREADSHEET_URL_VALID = "https://docs.google.com/spreadsheets/d/1iJqwMxLWOr1fDcYDNuV-CUp-1smHpYtYqvaTk14Nwek";
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: Indentação das variáveis diferente das funções

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        		showStepFour();
+        sizing.assertShowingCorrectStep(4);
+        
+        stepFour.assertShowingButtonsAfterFinish();
+        stepFour.assertCloseSizingImport();
+        
+        sizing.assertWithShowingError(false);
+    }
+	
+	@Test
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: Indentação diferente??

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    public void whenSubmitStepTwoWithoutFillRequiredFields_ShowError() {
+        MainPage mainPage = MainPage.produce(webDriver);
+        SizingImportUi sizing = mainPage.openSizingImport();
+        
+        sizing.
+            showStepOne().
+            withSpreadsheetUrl(SPREADSHEET_URL_VALID).
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: Indentação diferente

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        	submitStep().
+        	showStepTwo().
+            withRequiredFieldsWithoutValues().
+            submitStep();
+        
+        sizing.assertWithShowingError(true);
+        sizing.assertShowingCorrectStep(2);
+    }
+	
+	@Test
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: Não estou vendo o 'assert' desse teste

File: src/test/java/objective/taskboard/it/SizingImportUi.java [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+	public SizingImportUi(WebDriver driver) {
+		super(driver);
+	}
+
+    public static SizingImportUi open(WebDriver webDriver) {
+        return initElements(webDriver, SizingImportUi.class).open();
+    }
+
+    private SizingImportUi open() {
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportUi.java]: Indentação esquisita

File: src/test/java/objective/taskboard/it/SizingImportUi.java [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        waitVisibilityOfElement(sizing);
+        stepsNumbers = sizing.findElements(By.cssSelector(".steps__number[data-step]"));
+        return this;
+    }
+
+    public SizingStepOne showStepOne() {
+    	return new SizingStepOne();
+    }
+    
+    public SizingImportUi assertWithShowingError(Boolean showingError) {
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportUi.java]: Esse tipo de assert corre o risco de introduzir intermitência. O correto é "esperar" pela mensagem de erro

File: src/test/java/objective/taskboard/it/SizingImportUi.java [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+    private Boolean hasErrorMessage() {
+    	return sizing.findElements(By.id("errorMessage")).size() > 0;
+    }
+    
+    public void assertShowingCorrectStep(Integer stepNumber) {
+    	WebElement activeStepNumber = stepsNumbers.get(stepNumber - 1);
+    	waitVisibilityOfElement(activeStepNumber);
+    	
+    	List<WebElement> step = webDriver.findElements(By.cssSelector(".sizing__step.active[data-step='" + stepNumber + "'"));
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportUi.java]: Novamente, o certo é "esperar" que a classe esteja lá, e não verificar se está "agora"

File: src/test/java/objective/taskboard/it/SizingImportUi.java [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        }
+    }
+
+    abstract class SizingMiddleStep {
+
+    	protected Integer stepNumber;
+
+    	protected WebElement step;
+    	protected WebElement buttonBack;
+        protected WebElement buttonCancel;
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportUi.java]: Indentação esquisita

File: src/test/java/objective/taskboard/it/SizingImportUi.java [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+    	private void setDefaultFirstStepElements() {
+            step = sizing.findElement(By.cssSelector(".sizing__step[data-step='" + stepNumber + "']"));
+            waitVisibilityOfElement(step);
+            buttonCancel = step.findElement(By.cssSelector(".sizing__cancel"));
+            buttonSubmit = step.findElement(By.cssSelector(".sizing__submit"));
+            waitVisibilityOfElements(buttonCancel, buttonSubmit);
+    	}
+
+        public SizingStepOne assertHasProjects() {
+        	Assert.assertTrue(selectProject.getOptions().size() > 0);
```
Comment (taksan) [inline src/test/java/objective/taskboard/it/SizingImportUi.java]: Esse assert sofre do mesmo problema, não está esperando, está assumindo que será verdade na hora do assert

File: src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        if (!optionalCredential.isPresent())
+            return false;
+        
+        GoogleCredential credential = optionalCredential.get();
+
+        if (credential.getExpiresInSeconds() == null || credential.getExpiresInSeconds() < 60) {
+            try {
+                credential.refreshToken();
+
+            } catch (TokenResponseException e) {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java]: CRITICAL Either log or rethrow this exception. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+    @Override
+    public boolean verifyAuthorization() throws IOException {
+        Optional<GoogleCredential> optionalCredential = getCredential();
+        if (!optionalCredential.isPresent())
+            return false;
+        
+        GoogleCredential credential = optionalCredential.get();
+
+        if (credential.getExpiresInSeconds() == null || credential.getExpiresInSeconds() < 60) {
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java]: MAJOR Assign this magic number 60 to a well-named constant, and use the constant instead. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS109)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        sizing.
+            showStepOne().
+            submitStep();
+        
+        sizing.assertWithShowingError(true);
+        sizing.assertShowingCorrectStep(1);
+    }
+	
+    @Test
+    public void whenInputtingColumnValues_CheckIfTheValuesAreCorrect() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+            showStepOne().
+            withSpreadsheetUrl(SPREADSHEET_URL_INVALID).
+            submitStep();
+
+        sizing.assertWithShowingError(true);
+        sizing.assertShowingCorrectStep(1);
+    }
+
+    @Test
+    public void whenSubmitStepTwoWithRepeatedColumnField_ShowError() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+public class SizingImportIT extends AuthenticatedIntegrationTest {
+
+    final static String SPREADSHEET_URL_VALID = "https://docs.google.com/spreadsheets/d/1iJqwMxLWOr1fDcYDNuV-CUp-1smHpYtYqvaTk14Nwek";
+    final static String SPREADSHEET_URL_INVALID = "https://docs.google.com/spreadsheets/d/";
+
+    final static String LAST_COLUMN = "AB";
+    final static String INVALID_COLUMN = "AAAA";
+
+    @Test
+    public void whenFinishAllSteps_ImportSizingAndClose() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            showStepTwo().
+            withRepeatedColumns().
+            submitStep();
+
+        sizing.assertWithShowingError(true);
+        sizing.assertShowingCorrectStep(2);
+    }
+
+    @Test
+    public void whenSubmitStepTwoWithoutFillRequiredFields_ShowError() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+        stepFour.assertShowingButtonsAfterFinish();
+        sizing.assertWithShowingError(false);
+
+        stepFour.returnToTaskboard();
+        sizing.asserWithShowingSizing(false);
+    }
+
+    @Test
+    public void whenSubmitStepOneWithoutFillSpreadsheetUrl_ShowError() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

File: src/test/java/objective/taskboard/it/SizingImportIT.java [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            withColumnValue(1, LAST_COLUMN);
+
+        stepTwo.assertColumnValueEquals(0, INVALID_COLUMN, false);
+        stepTwo.assertColumnValueEquals(1, LAST_COLUMN, true);
+
+        sizing.assertWithShowingError(false);
+    }
+
+    @Test
+    public void whenSubmitStepOneWithInvalidSpreadsheetUrl_ShowError() {
```
Comment (objective-solutions-user) [inline src/test/java/objective/taskboard/it/SizingImportIT.java]: MAJOR Add at least one assertion to this test case. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2699)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 8 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 7 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
```
Comment (objective-solutions-user) [inline src/main/java/objective/taskboard/google/GoogleApiServiceHttp.java]: MAJOR Assign this magic number 60 to a well-named constant, and use the constant instead. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS109)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 2 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[CRITICAL]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

**Pull request merged**